### PR TITLE
Improve 404 redirect plugin

### LINKED ISSUE #353: Improve 404 redirect plugin
Apply the improvements made in TEXTman redirect plugin

See: https://github.com/joomlatools/textman/pull/513

### PR DESCRIPTION
**Pull request by allanpilarca**
Closes #353

- Handle platform running in subdirectory
- Dont include menu subpath when preparing the url to no-id

In this PR i kept the regex that will remove all the ids in the url `/xxx-` to `/`. No need to redirect twice to remove the category id then the article id.

### CODE REVIEW

**Comment by johanjanssens** [comment]
@ercanozkaya Please review

File: lib/plugins/system/404/404.php [hunk 1]
```diff
@@ -27,14 +27,21 @@ public static function handleError(JException $error)
 
         if ($app->get('sef') && (int) $error->getCode() == 404)
         {
+            $menu        = $app->getMenu()->getActive();
+            $menu_prefix = JRoute::_($menu->link, false);
+            $subpath     = preg_replace('#'.preg_quote(JURI::base(true)).'#', '', $menu_prefix, 1);
```
Comment (ercanozkaya) [inline lib/plugins/system/404/404.php]: having the replace=1 prefix is good but I'd still like to make sure we only replace the path from the beginning. Can you make the regexp start with a `^` please? Also if `JURI::base(true)` returns `/` this doesn't have to run right? We could add that for cleaner code as well.

Comment (allanpilarca) [reply lib/plugins/system/404/404.php]: @ercanozkaya the reason i implemented that way is to replace all `/id-` in one call and to lessen it to just one redirection. I can also apply the regex from PlgSystemTextman to have less code to maintain, what do you suggest? :)

Comment (ercanozkaya) [reply lib/plugins/system/404/404.php]: @allanpilarca Let's go with the code in Textman if possible. The two redirections look like overhead but they do support an edge case. For example `/movies/12-angry-men/3423-cast.html` would be redirected to `/movies/angry-men/cast.html`without it which breaks the link.

Comment (allanpilarca) [reply lib/plugins/system/404/404.php]: @ercanozkaya oki, i applied the regex code from TEXTman as you suggested.

**Pull request merged**